python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -815,8 +821,13 @@ def should_reconcile(
             return False
 
         if all(
-            will_be_materialized_for_freshness(candidate)
-            or parents_will_be_reconciled(asset_graph, candidate)
+            asset_graph.partition_maps_to_valid_parents(
```

Comment:
This should be in the `can_reconcile_candidate` above instead. As is, if we're in this situation, we'll hit the "else" case, which means that the reason that this asset is skipped will be registered as "ParentOutdated", which would be inaccurate.

It's arguable that we should carve out a separate reason for this case, but it's rare enough that for now it's ok to just leave blank.

Reply:
oo good catch

Reply:
Should this actually be in `parents_will_be_reconciled`? It seems like, from a logical perspective, finding out whether the parents are valid is part of finding out whether they're reconciled.

I'm imagining that eventually we'd want to get to a point where we just invoke the PartitionMapping once to find out what's up with the parent partitions: i.e. both what the parent partitions are whether any of the ones we need are non-existent, and then using that to answer the question of `parents_will_be_reconciled`.

Reply:
I actually think that it makes sense for this check to exist in `can_reconcile_candidate`. It's checking whether the partition itself has the capacity to be materialized, not the parents

So I think in this case, it doesn't matter that the parents will be reconciled if the candidate itself can't be materialized

Reply:
> It's checking whether the partition itself has the capacity to be materialized, not the parents

Whether the partition itself has the capacity to be materialized depends on the parents though, right?

The purpose of the `parents_will_be_reconciled` function is basically to make sure that the parents are in the state that's required to be able to materialize the child.

We might need to amend the name of `parents_will_be_reconciled` to something a little more general for this to make full sense.

Reply:
I agree that there is benefit in being able to only fetch upstream info from the partition mapping once. 

This is tricky in this case though since the `parents_will_be_reconciled` function is used to determine if parents are outdated, displaying a corresponding message in the UI if so. Adding this check in the `parents_will_be_reconciled` function would then display a "parent outdated" message in the UI when the upstreams are invalid, which would be inaccurate

Reply:
I'm happy to refactor the existing code to fetch the upstream info from the partition mapping once, and then propagate that information to `parents_will_be_reconciled`, but I would prefer to do that in the next PR in this stack since this change is slightly more involved (requires changing return types for methods on the asset graph)